common/workunit/workunit.cpp

```diff
@@ -9397,6 +9827,112 @@ IWUGraphStats *CLocalWorkUnit::updateStats(const char *graphName, StatisticCreat
+
+        const char *searchStr = strieq(processType, "Thor") ? THORMASTERLOGSEARCHSTR : ECLAGENTLOGSEARCHSTR;
+        if (strncmp(logName, searchStr, strlen(searchStr)) != 0) //example: mythor_thormaster 
+        {// This is an imported WU. Find processName from logName.
+            VStringBuffer afterStr("_%s", searchStr);
+            const char *ptr = logName.str();
+            const char *pptr = strstr(ptr, afterStr);
+            if (!pptr)
+                return false; //Should not happen
+            process.append(pptr - ptr, ptr);
+        }
+        else if (!path.isEmpty())
+        {// Find processName from path
+            path.setLength(path.length() - 1); //Remove the last char which is a PATHSEPCHAR
+            const char *ptr = strrchr(path.str(), PATHSEPCHAR);
+            if (!ptr)
+                return false; //Should not happen
+            process.set(ptr + 1);
+        }
+        else
+            return false; //Should not happen
+
+        getWUProcessLogInfo(processType, process, processLogs);
+    }
+    if (processLogs.ordinality() == 0)
+        return false;
+
+    return processLogs.item(0).getSingleLog();
+}
+
+unsigned CLocalWorkUnit::getNumberOfThorSlaves(const char *processName) const
+{
+    Owned<IPropertyTreeIterator> procs = getProcesses("Thor", processName);
+    if (procs->first())
+    {
+        unsigned numberOfThorSlaves = procs->query().getPropInt("@max", 0);
+        if (numberOfThorSlaves > 0)
+            return numberOfThorSlaves;
+    }
+
+    //Legacy WU
+    StringAttr clusterName(queryClusterName());
+    if (!clusterName.length())
+    { //Cluster name may not be set yet
+        IWARNLOG("Cluster name may not be set for workunit %s", queryWuid());
+        return 0;
+    }
+
+    Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
+    if (clusterInfo)
+        return clusterInfo->getNumberOfSlaveLogs();
+
+    IWARNLOG("Cannot find TargetClusterInfo for workunit %s", queryWuid());
+    return 0;
+}
+
+void CLocalWorkUnit::getWUProcessLogInfo(const char *processType, const char *processName, IArrayOf<IConstWUProcessLogInfo> & processLogs) const
+{
+    if (isEmptyString(processType))
+        return;
+
+    Owned<IPropertyTreeIterator> processInstances = getProcesses(processType, isEmptyString(processName) ? nullptr : processName);
+    ForEach (*processInstances)
+    {
+        StringBuffer logSpec;
+        IPropertyTree &proc = processInstances->query();
+        proc.getProp("@log", logSpec);
+
+        const char *processName = proc.queryName();
+        if (isEmptyString(processName) || logSpec.isEmpty())
+            continue;
+
+        unsigned numberOfThorSlaves = 0, processID = 0;
+        StringBuffer groupName;
+        if (strieq(processType, "Thor"))
+        {
+            getClusterThorGroupName(groupName, processName);
+            numberOfThorSlaves = getNumberOfThorSlaves(processName);
```

Comment:
getProcesses() is called (line 9903) and then you are iterating through them.
getNumberOfThorSlaves also calls getProcesses() but with the processName it got from the instance the iterator is on, so it is re-iterating to find a match of something it already has afaics?

shouldn't this pickup @max if present... and use the legacy code if not.
And since this is the only place getNumberOfThorSlaves is called afaics, then there's no point in getNumberOfThorSlaves calling getProceses().

And since getNumberOfThorSlaves isn't used other than in this context (but IConstWUProcessLogInfo::getNumberOfThorSlaves is), it would be better remove this method.

Reply:
Good catch.